File: z1. grep-shakespeare/grep.js
var	fs = require('fs'), path = require('path'), 
	shakespeare = 'pg100.txt',
	out = {},
	// output = 'nodejs.json',
	data = '',
	result = 'result.json'
	;

var strip = function(arr){
	return arr.map(function(item,iterator){
		return item.replace(/[^a-zA-Z ]/g, '');
	});
};

var add = function(words,cb){
	words.map(function(word,iterator){
		if (!out[word]) {
			out[word] = 1;
		} else {
			out[word]+=out[word];
		}		
	});
	cb();
};

var extract = function(file, cb){
	var rl = require('readline').createInterface({
		input: fs.createReadStream(path.join('./../data',file))
	});

	rl.on('line',function(line){
		cb(line);
	});

	rl.on('close',function(arr){
		console.log('done');
		console.log(out);
	});
};

var transform = function(line,cb){
	var words = strip(line.split(' '));

	cb(words);
};

extract(shakespeare,function(line){
	transform(line,function(words){
		add(words,function(){
			// console.log(out);			
		});
	});
});

File: z1. grep-shakespeare/grep.py
import re

shakeseare = './../data/pg100.txt'
out = dict()

def extract(file):
	for line in file:
		transform(line)
	file.close()

def transform(line):
	words = line.split()
	words = list(map(lambda x: strip(x), words))
	load(words)

def load(row):
	map(lambda x: add(x), row)

def add(word):
	if word in out:
		out[word] += 1
	else:
		out[word] = 1

def strip(word):
	return re.sub('[^A-Za-z]', '', word.lower())

def open_file(file):
	with open(file,'r') as f:
		extract(f)

open_file(shakeseare)
print(out)
